### PR TITLE
Search for preinstalled (read-only) extensions in XDG_DATA_DIRS

### PR DESCRIPTION
**Pull request by nazarewk**
I've changed the code to search through all `$XDG_DATA_DIRS` for extensions that are read-only.
Writes and edits are still done only to `$XDG_DATA_HOME`. 
It will help me a lot with packaging extensions on Nix as often it symlinks plugins/extensions into the package dir itself and/or augments binaries' standard environments extensively, having a single HOME directory would not do without additional tooling (like [`home-manager`](https://github.com/nix-community/home-manager) symlinking packages into HOME).

Scope:
- [x] preinstalled extensions are discovered and functional,
- [x] user-installed extensions are not affected by the change,
- [x] visually distinguish user-installed (primary directory) and preinstalled (secondary directories) extensions,
- [x] visually mark user-installed extensions that have their preinstalled counterparts with possible versions informations on both of those,
- [x] inform user about extension's location (for both primary and secondary directories),
- [x] disable `remove` button for preinstalled extensions,
- [x] ensure preinstalled extensions can be disabled,
- [x] updating preinstalled extension should simply create a fresh install into primary the directory
- [ ] add at least basic test coverage for above,

I also wonder if:
- it wouldn't be easier (especially for testing and mocking) for `_PATHS_CLASS` to be a collection of (cached?) properties?
- shouldn't we use `pathlib.Path` instead? would make most of paths handling code significantly shorter,

spun off #1307

### CODE REVIEW

**Comment by friday** [comment]
I need to think about this, in part because of the added complexity, and because I prefer if it's a simple default like ~/.local/share/ulauncher/extensions. That's what we've been telling people here if they ask where the extension dir is, and we direct users there is they can't install extensions due to network issues or similar (this extension url doesn't exist):

image

Also, locate_extension is confusing:
1. Why do you pass PATHS.EXTENSIONS_ALL in most places but PATHS.EXTENSIONS in [ExtensionController.py](https://github.com/Ulauncher/Ulauncher/pull/1319/files#diff-154426fce592377cfea20831957720ed5d19048431e7bf1b344b890966897217) and also default_first in this case.
2. For me at least EXTENSIONS_ALL doesn't contain the old extension path (because XDG_DATA_DIRS doesn't include XDG_DATA_HOME), so I think this would be breaking?

_PATHS_CLASS could most likely be improved yes. I just tried to make something useful that worked for old python versions. But since then we have dropped 3.6 support and we even have our own data class for compatibility reasons. I don't understand what you're suggesting or why though.

os.scandir is supposed to be faster. Probably not noticably so (I never tested), but before there wasn't really any complexity cost to using os.scandir instead of pathlib. But if you want to switch and make a second PR with that change I could understand the difference better.

File: ulauncher/config.py
Comment (troycurtisjr) [inline ulauncher/config.py:20]: I know that it means more code changes, but if we want to make this change, this variable should be more descriptive. As it is, the reader needs read the other variable, `EXTENSIONS_ALL`, to get a hint about what this variable isn't, but even then it isn't very descriptive other than knowing that it "doesn't contain all the extensions". Instead, this should be `MUTABLE_EXTENSIONS` or something like that.

File: ulauncher/modes/extensions/extension_finder.py
Comment (troycurtisjr) [inline ulauncher/modes/extensions/extension_finder.py:13]: Unless I'm missing it, `default` is never given by any caller. Additionally, the parameter `default_first` makes it sound like the default given will be used first, when instead the first path in the configuration is used, and the default (if it were ever set) would be returned directly.

**Review by troycurtisjr — CHANGES_REQUESTED** [review]
I personally like the concept, as I have also thought about ways we could have system packaged extension. It does add some complexity though, and while this PR is a good starting point, it probably doesn't handle enough of the situations yet. 

@friday brought up one of the cases around where we are pointing users to when errors occur, but we also have to consider the proper behavior when:

1. A user tries to delete an extension they don't have permission to delete
2. Should we display something if an extension is installed in multiple places?
3. How do we behave when a user requests an update to an extension they can't write to?

File: ulauncher/modes/extensions/ExtensionDownloader.py
Comment (nazarewk) [reply ulauncher/modes/extensions/ExtensionDownloader.py:37]: yeah, would make it a lot clearer

File: ulauncher/modes/extensions/ExtensionDownloader.py
Comment (nazarewk) [reply ulauncher/modes/extensions/ExtensionDownloader.py:39]: I don't know the internals well enough, so I fallback to doing nothing like the original

File: ulauncher/modes/extensions/ExtensionRemote.py
Comment (nazarewk) [reply ulauncher/modes/extensions/ExtensionRemote.py:84]: Yeah, I think having a `get_mutable_extension_dir()` would make it a lot clearer

**Comment by nazarewk** [comment]
> I need to think about this, in part because of the added complexity, and because I prefer if it's a simple default like ~/.local/share/ulauncher/extensions. That's what we've been telling people here if they ask where the extension dir is, and we direct users there is they can't install extensions due to network issues or similar (this extension url doesn't exist):

I do agree it is added complexity, but:
1. existing users won't know a difference (the primary directory will work exactly the same)
2. package maintainers will be able to package extensions,

> Also, locate_extension is confusing:
> 
>     1. Why do you pass PATHS.EXTENSIONS_ALL in most places but PATHS.EXTENSIONS in [ExtensionController.py](https://github.com/Ulauncher/Ulauncher/pull/1319/files#diff-154426fce592377cfea20831957720ed5d19048431e7bf1b344b890966897217) and also default_first in this case.
> 
>     2. For me at least EXTENSIONS_ALL doesn't contain the old extension path (because XDG_DATA_DIRS doesn't include XDG_DATA_HOME), so I think this would be breaking?

added some refactoring according to @troycurtisjr comments, should be a lot clearer now. During refactoring it made me think `ULAUNCHER_DATA_DIR` is a confusing name, it suggests XDG counterpart of DATA_DIR, but in reality it is an assets/sources directory.

> _PATHS_CLASS could most likely be improved yes. I just tried to make something useful that worked for old python versions. But since then we have dropped 3.6 support and we even have our own data class for compatibility reasons. I don't understand what you're suggesting or why though.

it just seems like an unusual construct, but I don't have any strong opinions on it

> os.scandir is supposed to be faster. Probably not noticably so (I never tested), but before there wasn't really any complexity cost to using os.scandir instead of pathlib. But if you want to switch and make a second PR with that change I could understand the difference better.

I don't know about speed, but I am quite sure the code itself would be a lot shorter/clearer in a lot of places if we started with `Path` objects in `config.py` and followed through. I'm not against using something faster where needed (whole `os` module should be compatible with Path anyway?)

----

> @friday brought up one of the cases around where we are pointing users to when errors occur, but we also have to consider the proper behavior when:
>   1. A user tries to delete an extension they don't have permission to delete
>   2. Should we display something if an extension is installed in multiple places?
>   3. How do we behave when a user requests an update to an extension they can't write to?

Yeah, I agree there is a lot more to think about (I just quickly checked if it worked at all):
- any extension located outside the primary user directory should most likely be
	- marked as system-provided and/or preinstalled extension,
	- 2) path (also for user extensions) should be displayed somewhere,
	- 1) non-removable (button greyed out),
	- 1) should be disableable, I don't know yet how is it implemented
- 2) I think the only case we should inform user about would be telling about user-provided extension also existing in system config,
- 3) updating/modifying such extension should simply write to primary user directory,

**Comment by nazarewk** [comment]
I picked up the `SYSTEM_*` vs `USER_*` themes naming and took liberty of normalizing variable names where it made sense.

File: ulauncher/modes/extensions/ExtensionDownloader.py
Comment (friday) [reply ulauncher/modes/extensions/ExtensionDownloader.py:39]: We probably should prevent this from being called in the first place if it's in the system path, so this shouldn't be called then? And in that case maybe we can just log a warning here?

**Comment by friday** [comment]
Really nice work! With the refactoring the picture got a bit clearer for me at  least (too busy to do a full review of this still). I think it is at least in our interest to support something close to this, but I need time before I can review this.

About ULAUNCHER_DATA_DIR, feel free to rename that. I added a comment to myself not to do it because I didn't want to break your changes :D

**Comment by nazarewk** [comment]
I've updated the UI testing functionality along the way:

<details><summary>preinstalled extension</summary>
<p>

screenshot-2023-11-22_13-19-49

</p>
</details> 

<details><summary>user installed extension</summary>
<p>

screenshot-2023-11-22_13-20-13

</p>
</details> 

<details><summary>user extension installed on top of preinstalled extension</summary>
<p>

screenshot-2023-11-22_13-20-49

</p>
</details> 

Things I noted during implementation:
- preferences: `remove` button did not reload list of extensions after removing them,
- extensions were not registered anywhere except during the download or update, so preinstalled extensions were in some weird half-installed state,
	- I have added extensions registration and starting during both instances of extensions discoveri (`run_all()` and `get_extensions()`) to solve it

Suggestions:
- `manifest.json` could be augmented with `repository_url` field for checking updates/updating preinstalled extensions
- there could be `.extension.json` file to be merged into `ExtensionRecord` (could be used instead of `manifest.json`), it would help a lot with preinstalled extensions registration.

**Comment by nazarewk** [comment]
i've implemented `.extension-record.json` handling to provide the default configuration alongside sources/inside installation dir.

**Comment by friday** [comment]
I had a look at it and pushed a couple of small commits with things I thought was easier to fix instead of commenting.

According to https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables
> If $XDG_DATA_DIRS is either not set or empty, a value equal to /usr/local/share/:/usr/share/ should be used.

This is similar to why we fallback for the other variables like "XDG_DATA_HOME", so I rewrote it to do that here as well.

And I removed the PREINSTALLED_EXTENSIONS_DIRS since that was not used outside of the config module any more now.

I am still reviewing so I will comment more in a little while.

**Comment by friday** [comment]
Pushed two more changes:
1. I don't think loading the extension page should try to start extensions that are not running, so I changed it so that it doesn't do that, but instead repurposed the "Reload the list" link so it reloads extensions now.
2. The code that loads the extensions internally runs validations and can throw, so I moved it to the try/except and skip the validation unless the extensions are disabled

**Comment by friday** [comment]
Really great work! I  did some testing and fixes + one variable refactoring commit. Hopefully you agree. From Ulaunchers perspective we have one path that we do manage. The rest we shouldn't manage, and we don't know/need to know if those are preinstalled, system paths or users home dir paths.

XDG_DATA_DIRS usually contain `/usr/local/share/` and `/usr/share/` as mentioned, but for me it also contains `~/.local/share/flatpak/exports/share` and `/var/lib/flatpak/exports/share`. This is something I think we'd need to consider if we want. I think it doesn't quite make sense, because if we do choose to support flatpak in the future we'd either want to use the same `USER_EXTENSIONS_DIR` path as now or use a path inside of `~/.local/share/flatpak/exports/share` as the USER_EXTENSIONS_DIR only for the flatpak version and not load those for the non-flatpak version (and we'd never ues `/var/lib/flatpak/exports/share`). At least that's what I think would make more sense?

The only other part I'm not following with is the .extension-record.json thing? Is this the same info that's stored in f"{PATHS.CONFIG}/extensions.json" now? Why store (duplicate?) that in the extension path? What about extension paths that are not mutable?

Other than that I think this looks great now. You left a couple of TODOs that I suppose you wanted to handle? (it doesn't need to be perfect though)

> `manifest.json` could be augmented with `repository_url` field for checking updates/updating preinstalled extensions

:+1: This is something I have had on my mental list to do, because I install my own extensions from the local path like,  `~/src/personal/ulauncher-clipboard` and I would want that to replace the installed extension for https://github.com/friday/ulauncher-clipboard

> there could be `.extension.json` file to be merged into `ExtensionRecord` (could be used instead of `manifest.json`), it would help a lot with preinstalled extensions registration.

I agree on this too, although it hasn't been a high prio for me.

**Comment by friday** [comment]
I just realized you are only reading from .extension-record.json, so along with your comment it finally clicked. I guess this is something that makes sense for Nix mostly?

**Comment by friday** [comment]
Maybe I was thinking wrong here cb16dbd141eee7a868172c31602a8a90c7abbca8.

runner.run() calls extension_finder.locate(), so there's no need to do that separately, but the exception handling is needed around runner.run() instead, so this part should be right now (Igot a crash from the extension trying to start itself when I removed it before).

But the extra `was_running` condition maybe shouldn't be there? I thought it was strange that removing a stopped extension would start it, but I guess you have a corrupt one that you uninstall and there is a working one in the XDG_DATA_DIRS paths it could make sense (especially since it will start anyway if you reload unless it's diasbled)?

**Comment by nazarewk** [comment]
> 1. XDG_DATA_DIRS usually contain `/usr/local/share/` and `/usr/share/` as mentioned, but for me it also contains `~/.local/share/flatpak/exports/share` and `/var/lib/flatpak/exports/share`. This is something I think we'd need to consider if we want. I think it doesn't quite make sense, because if we do choose to support flatpak in the future we'd either want to use the same `USER_EXTENSIONS_DIR` path as now or use a path inside of `~/.local/share/flatpak/exports/share` as the USER_EXTENSIONS_DIR only for the flatpak version and not load those for the non-flatpak version (and we'd never ues `/var/lib/flatpak/exports/share`). At least that's what I think would make more sense? I'm not sure I'm in the right here though, and I'm also not sure there is any alternative to XDG_DATA_DIRS?

We should not concern ourselves with too many `XDG_DATA_DIRS` or whether it contains entries we are not interested in at all (this is what `/ulauncher` application subdirectory is for). This list can get VERY long and it's normal:
```
> echo $XDG_DATA_DIRS | tr ':' '\n'
/nix/store/k5nbwh8filcsb4qnfad27gywaqmccbyy-python3.11-ulauncher-v6/share
/nix/store/drdhg1pghyl0dvmngkadv6viv8jrm0zc-gsettings-desktop-schemas-44.0/share/gsettings-schemas/gsettings-desktop-schemas-44.0
/nix/store/6qp789r70ka5ahxfzrr9l3h6szclsgsr-gtk+3-3.24.38/share/gsettings-schemas/gtk+3-3.24.38
/nix/store/b2x5i6bqhk3hnqpz5f9bhs2rx89wxzl9-sway-1.8.1/share
/nix/store/drdhg1pghyl0dvmngkadv6viv8jrm0zc-gsettings-desktop-schemas-44.0/share/gsettings-schemas/gsettings-desktop-schemas-44.0
/nix/store/6qp789r70ka5ahxfzrr9l3h6szclsgsr-gtk+3-3.24.38/share/gsettings-schemas/gtk+3-3.24.38
/nix/store/n0aj5bvxfb95znkf49p6v23g0qrnp4pp-desktops/share
/home/kdn/.nix-profile/share
/nix/profile/share
/home/kdn/.local/state/nix/profile/share
/etc/profiles/per-user/kdn/share
/nix/var/nix/profiles/default/share
/run/current-system/sw/share
```

**Comment by nazarewk** [comment]
> Maybe I was thinking wrong here [cb16dbd](https://github.com/Ulauncher/Ulauncher/commit/cb16dbd141eee7a868172c31602a8a90c7abbca8).
> 
> runner.run() calls extension_finder.locate(), so there's no need to do that separately, but the exception handling is needed around runner.run() instead, so this part should be right now (Igot a crash from the extension trying to start itself when I removed it before).
> 
> But the extra `was_running` condition maybe shouldn't be there? I thought it was strange that removing a stopped extension would start it, but I guess you have a corrupt one that you uninstall and there is a working one in the XDG_DATA_DIRS paths it could make sense (especially since it will start anyway if you reload unless it's diasbled)?

You are right, I should not run extension from another location which was not running before removal.

**Comment by nazarewk** [comment]
> I just realized you are only reading from .extension-record.json, so along with your comment it finally clicked. I guess this is something that makes sense for Nix mostly?

Yeah, `.extension-record.json` is meant to supersede that `repository_url` at `manifest.json` suggestion. I have reused the class name of the objects it is providing defaults for.

It is not specific to any single package manager/installation method. It is supposed to be a general purpose mechanism providing `ulauncher` with extension details it is unable to discover outside the *install from git repository* flow.

> 👍 This is something I have had on my mental list to do, because I install my own extensions from the local path like, ~/src/personal/ulauncher-clipboard and I would want that to replace the installed extension for [friday/ulauncher-clipboard](https://github.com/friday/ulauncher-clipboard)

You have found one use case yourself :)

> I think I would prefer if you reverted that for now and handle it in a separate PR so this one doesn't grow even more :D

I tried quickly dropping the commit into separate PR, but it seems to be heavily embedded into the flow and would take a while to carve out, so I would prefer to merge it here and refine later on.

> And if you have more small fixes and formatting changes I much prefer several small PRs than one big (separate commits in one PR is also good, but you can't approve those separately).

I don't think I want to make any more changes before merging.

**Comment by friday** [comment]
> Yeah, .extension-record.json is meant to supersede that repository_url at manifest.json suggestion. I have reused the class name of the objects it is providing defaults for.

If we support repository_url do you still need/want . extension-record.json?

**Comment by nazarewk** [comment]
> > Yeah, .extension-record.json is meant to supersede that repository_url at manifest.json suggestion. I have reused the class name of the objects it is providing defaults for.
> 
> If we support repository_url do you still need/want `.extension-record.json`?

`repository_url` would simply give you something to link to without any version information which `ExtensionRecord` stores and that could be provided externally during installation/packaging step.

**Comment by friday** [comment]
I pushed another very small UI change. I will merge it now and we can discuss the `.extension-record.json` file further if I come up with a counter-proposal.

**Pull request merged**
